state-sync/storage-service/server/src/lib.rs

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use diem_infallible::RwLock;
+use diem_types::{epoch_change::EpochChangeProof, transaction::TransactionListWithProof};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use storage_interface::DbReaderWriter;
+use storage_service_types::{
+    DataSummary, EpochEndingLedgerInfoRequest, ProtocolMetadata, ServerProtocolVersion,
+    StorageServerSummary, StorageServiceError, StorageServiceRequest, StorageServiceResponse,
+    TransactionsWithProofRequest,
+};
+use thiserror::Error;
+
+#[cfg(test)]
+mod tests;
+
+// TODO(joshlind): make these configurable.
+/// Storage server constants.
+pub const MAX_TRANSACTION_CHUNK_SIZE: u64 = 1000;
+pub const STORAGE_SERVER_VERSION: u64 = 1;
+
+#[derive(Clone, Debug, Deserialize, Error, PartialEq, Serialize)]
+pub enum Error {
+    #[error("Storage error encountered: {0}")]
+    StorageErrorEncountered(String),
+    #[error("Unexpected error encountered: {0}")]
+    UnexpectedErrorEncountered(String),
+}
+
+/// The server-side implementation of the storage service. This provides all the
+/// functionality required to handle storage service requests (i.e., from clients).
+pub struct StorageServiceServer<T> {
+    storage: T,
+}
+
+impl<T: StorageReaderInterface> StorageServiceServer<T> {
+    pub fn new(storage: T) -> Self {
+        Self { storage }
+    }
+
+    pub fn handle_request(
+        &self,
+        request: StorageServiceRequest,
+    ) -> Result<StorageServiceResponse, Error> {
+        let response = match request {
+            StorageServiceRequest::GetEpochEndingLedgerInfos(request) => {
+                self.get_epoch_ending_ledger_infos(request)
+            }
+            StorageServiceRequest::GetServerProtocolVersion => self.get_server_protocol_version(),
+            StorageServiceRequest::GetStorageServerSummary => self.get_storage_server_summary(),
+            StorageServiceRequest::GetTransactionsWithProof(request) => {
+                self.get_transactions_with_proof(request)
+            }
+        };
+
+        // If any requests resulted in an unexpected error, return an InternalStorageError to the
+        // client and log the actual error.
+        if let Err(_error) = response {
+            // TODO(joshlind): add logging support to this library so we can log _error
+            Ok(StorageServiceResponse::StorageServiceError(
+                StorageServiceError::InternalError,
+            ))
```

Comment:
Interesting, you convert errors to responses that are Ok responses.  Maybe that's a little counterintuitive to me.  But, I guess you're putting the true "service" facade around it?  Like what is an error that would be thrown from here?

Reply:
See my comment to Philip above. This will need some more attention 😄